src/cached_object_store.rs

```diff
@@ -0,0 +1,1067 @@
+        match entry.read_head().await {
+            Ok(Some((meta, _))) => Ok(meta),
+            _ => {
+                let result = self
+                    .object_store
+                    .get_opts(
+                        location,
+                        GetOptions {
+                            range: None,
+                            head: true,
+                            ..Default::default()
+                        },
+                    )
+                    .await?;
+                let meta = result.meta.clone();
+                self.save_result(result).await.ok();
+                Ok(meta)
+            }
+        }
+    }
+
+    pub async fn cached_get_opts(
+        &self,
+        location: &Path,
+        opts: GetOptions,
+    ) -> object_store::Result<GetResult> {
+        let (meta, attributes) = self.maybe_prefetch_range(location, &opts.range).await?;
+        let range = self.canonicalize_range(opts.range, meta.size)?;
+        let parts = self.split_range_into_parts(range.clone());
+
+        // read parts, and concatenate them into a single stream. please note that some of these part may not be cached,
+        // we'll still fallback to the object store to get the missing parts.
+        let futures = parts
+            .into_iter()
+            .map(|(part_id, range_in_part)| self.read_part(location, part_id, range_in_part))
+            .collect::<Vec<_>>();
+        let result_stream = stream::iter(futures).then(|fut| fut).boxed();
+
+        Ok(GetResult {
+            meta,
+            range,
+            attributes,
+            payload: GetResultPayload::Stream(result_stream),
+        })
+    }
+
+    // TODO: implement the put with cache here
+    #[allow(unused)]
+    async fn cached_put_opts(
+        &self,
+        location: &Path,
+        payload: object_store::PutPayload,
+        opts: object_store::PutOptions,
+    ) -> object_store::Result<PutResult> {
+        self.object_store.put_opts(location, payload, opts).await
+    }
+
+    // if an object is not cached before, maybe_prefetch_range will try to prefetch the object from the
+    // object store and save the parts into the local disk cache. the prefetching is helpful to reduce the
+    // number of GET requests to the object store, it'll try to aggregate the parts among the range into a
+    // single GET request, and save the related parts into local disks together.
+    // when it sends GET requests to the object store, the range is expected to be ALIGNED with the part
+    // size.
+    async fn maybe_prefetch_range(
+        &self,
+        location: &Path,
+        range: &Option<GetRange>,
+    ) -> object_store::Result<(ObjectMeta, Attributes)> {
+        let entry = self.cache_storage.entry(location, self.part_size);
+        match entry.read_head().await {
+            Ok(Some((meta, attrs))) => return Ok((meta, attrs)),
+            Ok(None) => {}
+            Err(_) => {
+                // TODO: add a warning log
```

Comment:
i'm planning to add the warning logs in a follow up pr, this will be a trivial change, i'd go through the code again and cover the needed logs in a unified format.